### ISSUE TITLE
Alpha parameter doesn't work in bar3d

### ISSUE THREAD

**Issue by ngonthier** [matplotlib/matplotlib#9559]
### Bug report

I have the same problem than on this [Stackoverflow webpage](https://stackoverflow.com/questions/43670867/matplotlib-alpha-transparency-parameter-doesnt-work) : 

**Code for reproduction**

```python
from mpl_toolkits.mplot3d import Axes3D
import numpy as np
import matplotlib.pyplot as plt

def main():
	plt.ion()
	rho = np.random.random((4,4))
	xpos = np.arange(0,4,1)
	ypos = np.arange(0,4,1)
	xpos, ypos = np.meshgrid(xpos, ypos)
	xpos = xpos.flatten()
	ypos = ypos.flatten()
	zpos = np.zeros(4*4)
	dx = 0.5 * np.ones_like(zpos)
	dy = dx.copy()
	dz = rho.flatten()
	alpha = 0.5
	fig = plt.figure()
	ax = fig.gca(projection='3d')
	ax.bar3d(xpos,ypos,zpos, dx, dy, dz, alpha=alpha, linewidth=0)
	plt.show()
	input("Press Enter to continue...")

if __name__ == '__main__':
	main()

```

**Actual outcome**

The bar on the plot are not transparent at all but they should be.

**Matplotlib version**

  * Operating system: Ubuntu 16.04
  * Matplotlib version: 2.0.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: 3.6

Matplotlib have been installerd from conda.

**Comment by WeatherGod** [comment]
I wonder if this is backend-dependent. I can't reproduce this problem using
TkAgg (I don't have Qt5 installed on this machine)

On Tue, Oct 24, 2017 at 11:36 AM, Gonthier Nicolas <notifications@github.com

**Comment by anntzer** [comment]
I can reproduce on tkagg and qt5agg (they both use the same renderer anyways so it'd have been a really weird bug if the result was different).  This appears to be a regression somewhere between 1.5.0 and 2.0.0.

**Comment by WeatherGod** [comment]
Well, here is my image using v2.0.0 using TkAgg:

![figure_1](https://user-images.githubusercontent.com/291576/31964991-e36f1068-b8d3-11e7-8a74-4cfe06ed6ed3.png)

And that is the same as the image I get using v2.0.1 and v2.1.0...

**Comment by anntzer** [comment]
```
$ MPLBACKEND=tkagg MPLCONFIGDIR=/tmp/mpld python -c 'from mpl_toolkits import mplot3d; print(mplot3d)'
<module 'mpl_toolkits.mplot3d' from '/home/antony/src/extern/matplotlib/lib/mpl_toolkits/mplot3d/__init__.py'>
```
This is my source tree where I have 2.0 checked out.

Can you confirm on your side?

**Comment by WeatherGod** [comment]
For me, matplotlib and mpl_toolkits/mplot3d match up. They were installed via conda for the v2.0.0, v2.0.1, and v2.1.0 versions I checked.

This for the v2.0.0 install I have:
```
(hurricane) bash-4.2$ MPLBACKEND=tkagg MPLCONFIGDIR=/tmp/mpld python -c 'from mpl_toolkits import mplot3d; print(mplot3d)'
<module 'mpl_toolkits.mplot3d' from '/home/broot/scratch/miniconda/envs/hurricane/lib/python2.7/site-packages/mpl_toolkits/mplot3d/__init__.pyc'>
(hurricane) bash-4.2$ MPLBACKEND=tkagg MPLCONFIGDIR=/tmp/mpld python -c 'import matplotlib; print(matplotlib)'
<module 'matplotlib' from '/home/broot/scratch/miniconda/envs/hurricane/lib/python2.7/site-packages/matplotlib/__init__.pyc'>
(hurricane) bash-4.2$ ls -l /home/broot/scratch/miniconda/envs/hurricane/lib/python2.7/site-packages/mpl_toolkits/mplot3d/
```

And this for the v2.1.0 install:
```
bash-4.2$ MPLBACKEND=tkagg MPLCONFIGDIR=/tmp/mpld python -c 'from mpl_toolkits import mplot3d; print(mplot3d)'
<module 'mpl_toolkits.mplot3d' from '/home/broot/scratch/miniconda/lib/python2.7/site-packages/mpl_toolkits/mplot3d/__init__.pyc'>
bash-4.2$ MPLBACKEND=tkagg MPLCONFIGDIR=/tmp/mpld python -c 'import matplotlib; print(matplotlib)'
<module 'matplotlib' from '/home/broot/scratch/miniconda/lib/python2.7/site-packages/matplotlib/__init__.pyc'>
```

**Comment by anntzer** [comment]
Ah, of course, Py2 (I can confirm it works there too).
So it's a Py3-only issue.

**Comment by WeatherGod** [comment]
The only possibility I can think of there is dictionary order... but we addressed those issues a long time ago, and usually that had to deal with the order of colliding properties...

Maybe the alpha of the default color is overriding the user's alpha (doubtful...)?

**Comment by jsh9** [comment]
And then I inserted this line 

```
kwargs = {k: kwargs[k] for k in reversed(list(kwargs.keys()))}
```

immediately before this location (https://github.com/matplotlib/matplotlib/blob/master/lib/mpl_toolkits/mplot3d/art3d.py#L542), which simply created a new dictionary reversing the key order (so that `alpha` was the first key), and I was able to get semi-transparent 3D bars in Python 3.

Admittedly this is only a temporary hack. The matplotlib developers should be able to come up with a rigorous patch.

**Comment by anntzer** [comment]
Alternatively, something like
```
diff --git a/lib/mpl_toolkits/mplot3d/axes3d.py b/lib/mpl_toolkits/mplot3d/axes3d.py
index 1bd3cf7c0..2ad640943 100644
--- a/lib/mpl_toolkits/mplot3d/axes3d.py
+++ b/lib/mpl_toolkits/mplot3d/axes3d.py
@@ -2529,8 +2529,8 @@ class Axes3D(Axes):
 
         col = art3d.Poly3DCollection(polys,
                                      zsort=zsort,
-                                     facecolor=sfacecolors,
-                                     *args, **kwargs)
+                                     *args, **kwargs,
+                                     facecolor=sfacecolors)
         self.add_collection(col)
 
         self.auto_scale_xyz((minx, maxx), (miny, maxy), (minz, maxz), had_data)
```
also fixes the issue.  But note that both patches only work on Py3.6 as they rely on kwargs conserving kwargs order.

But it's really a problem of Poly3DCollection only updating the facecolors with the alpha if the alpha is already set when the colors are given (and the luck that on Py2, "alpha" comes before "facecolor" in dict iteration order...).  Something like
```
p = ax.bar3d(0, 0, 0, 1, 1, 1); p.set_alpha(.5)
```
also fails to apply the alpha on Py2.

**Comment by rth** [comment]
To follow up on the previous comment,  `bar3d` passes the `alpha` to `set_alpha`,
https://github.com/matplotlib/matplotlib/blob/8168210211c598f32f683084460b104a3fe2255c/lib/mpl_toolkits/mplot3d/axes3d.py#L2383-L2384


For me with Python 3, `p.set_alpha` (where `p` is the `Poly3DCollection` object) also has no effect, so the issue is with this method IMO.

Internally, `set_alpha` mostly consists of,
https://github.com/matplotlib/matplotlib/blob/4536165d1787ff980065c551a0d881df36079bf6/lib/mpl_toolkits/mplot3d/art3d.py#L694-L704
 
So using,
```py
p._facecolors3d = matplotlib.colors.to_rgba_array(
                          p._facecolors3d, p._alpha)
p._edgecolors3d = matplotlib.colors.to_rgba_array(
                          p._edgecolors3d, p._alpha)
```
and combining it with the fix proposed in https://github.com/matplotlib/matplotlib/issues/4067#issuecomment-357794003 (I was getting the corresponding error anyway),
```py
p._facecolors2d = p._facecolors3d
p._edgecolors2d = p._edgecolors3d
```
appears to make transparency work as expected.

**Comment by ImportanceOfBeingErnest** [comment]
Could be the same issue as #10237 for which there were already 2 attempted fixes, but none made it through yet. The one in #10487 looks promising and should just be a matter of getting merged?

**Comment by rth** [comment]
Yes, I think this is a duplicate of #10237. The issue is ultimately with `Poly3DCollection` from what I saw. It's docstring states,

>  Note that this class does a bit of magic with the _facecolors and _edgecolors properties. 

and some of that magic must not work as expected.

**Comment by ksunden** [comment]
Appears to have been fixed by #13573 (at least that is what closed #10237, which this appears to be a duplicate of)
